allennlp/training/trainer.py

```diff
@@ -948,13 +961,11 @@ def from_params(cls,
         patience = params.pop_int("patience", None)
         validation_metric = params.pop("validation_metric", "-loss")
         num_epochs = params.pop_int("num_epochs", 20)
-        cuda_device = params.pop_int("cuda_device", -1)
+        cuda_device = params.pop("cuda_device", -1)
```

Comment:
I don't think this will play well with environment variable substitution.

That is, if someone is parameterizing their config file with a `CUDA_DEVICE` environment variable, it will get read in as a string (like all environment variables), and then this will pop it as a string, which I think will break code that relies on it.

You probably will need to write an ugly helper function that takes a `Union[str, int, List[int]]` and parses the string to `int` or `List[int]` as appropriate

Reply:
Hmm - I think that should be the responsibility of user to do in Jsonnet (presuming that is possible in Jsonnet).

Reply:
> You probably will need to write an ugly helper function that takes a Union[str, int, List[int]] and parses the string to int or List[int] as appropriate

That isn't even ugly, I think -- shouldn't this be the default approach to processing environment variables? I.e. pick the right alternative out of all acceptable conversion targets and complain if none work or the choice is ambiguous. I'll see if I can come up with something reasonably generic...

Reply:
> I think that should be the responsibility of user to do
I guess I don't quite understand then -- if there's a way to override the configuration parameter via an environment variable (a type-less string value) for a specific run, how can the creator of the config (or the component that the config is intended for) manage the type inference? There is no schema for the configuration, is there?

Reply:
@DeNeutoy  the whole point of using `pop_int` here is that we don't think it's the user's responsibility (and how could it be, if they're using our trainer?)

Reply:
@MaxMotovilov I don't think it's the right approach in general, I think the current `pop_int` + `pop_bool` + `pop_float` is a good approach ("explicit is better than implicit"), it's just in this case we're allowing an int or a list, which is unusual. the function you write should fail if it gets something that can't be parsed as a list or an int

Reply:
It should be explicit anyway -- the alternatives have to be specified at some point -- which, in absence of a schema, is the point of use. In this particular case however we are not even dealing with a union: a general case of `List[int]` with a special case of one-element list equivalent to an int. The default (which is represented by -1 mostly for legacy reasons, I take it) really is a null value or an empty list: no GPU available.

So, how about adding a `params.pop_int_list()` (my ancient C++ roots cry for `params.pop[List[int]]()` but this is not to be) that handles `/,\s*/`-separated values and a default?

Reply:
It's your codebase -- you get to call'em :-)